File: infra/scripts/sync_dots.ts
import { join } from "https://deno.land/std@0.178.0/path/mod.ts";

const home = Deno.env.get("HOME");
if (home === undefined) {
  console.error(
    "No $HOME variable found, this script only works on Linux environment.",
  );
  Deno.exit(1);
}

const from_home = (...path: string[]) => {
  return join(home, ...path);
};

const pending_files = [
  from_home(".vimrc"),
  from_home(".bashrc"),
  from_home(".tmux.conf"),
];

const servers = Deno.args.length > 0 ? Deno.args : [
  // 5950x
  "shinx",
  "luxio",
  "manectric",
  "minun",

  // unmatched stable
  "larvesta",
  "reshiram",
  "incineroar",
  "talonflame",

  // unmatched unstable
  "volcanion",
  "torracat",
  "oricorio",
];

const rsync = async (remote: string, ...path: string[]) => {
  const handle = new Deno.Command("rsync", {
    args: [
      "-azvhP",
      ...path,
      `${remote}:~/`
    ]
  }).spawn();
  const status = await handle.status;
  if (!status.success) {
    console.error(`fail to sync file to ${remote}`)
  }
}

for (const srv of servers) {
  console.log(`==> Sending file to ${srv}`)
  rsync(srv, ...pending_files)
}

File: infra/bin/syncdots.ml
let home = Sys.getenv "HOME"

let resolve_dots (dot : string) =
  let is_symlink (stat : Unix.stats) = stat.Unix.st_kind = Unix.S_LNK in
  let fullpath = Filename.concat home dot in
  let stat = Unix.lstat fullpath in
  if is_symlink stat then Unix.readlink fullpath else fullpath

let servers =
  if Array.length Sys.argv > 1 then Sys.argv
  else
    [|
      (* 5950x *)
      "shinx";
      "luxio";
      "manectric";
      "minun";
      (* Unmatched *)
      "larvesta";
      "reshiram";
      "incineroar";
      "talonflame";
      "volcanion";
      "torracat";
      "oricorio";
    |]

let dotfiles =
  [ ".vimrc"; ".bashrc"; ".tmux.conf" ]
  |> List.map (fun dot -> resolve_dots dot)

let wait_process proc =
  let _, status = Unix.process_full_pid proc |> Unix.waitpid [] in
  match status with
  | Unix.WEXITED 0 -> Ok ()
  | _ ->
      let _, _, stderr = proc in
      Error (In_channel.input_all stderr)

(* Perform a `rsync -azvhP $files $remote:~/` command execution *)
let rsync ~(remote : string) ~(files : string list) =
  let remote_target = Printf.sprintf "%s:~/" remote in
  let rsync_arg =
    ("rsync" :: "-azvhP" :: files) @ [ remote_target ] |> Array.of_list
  in
  let handle_rsync proc =
    match wait_process proc with
    | Ok () -> Format.printf "==> %s DONE\n" remote
    | Error err ->
        Format.printf "Fail to sync file to %s\n" remote;
        Format.printf "--------------- stderr --------------\n";
        Format.printf "%s\n\n" err
  in
  let proc =
    Unix.open_process_args_full "rsync" rsync_arg (Unix.environment ())
  in
  Out_channel.flush_all ();
  handle_rsync proc

let () = servers |> Array.iter (fun srv -> rsync ~remote:srv ~files:dotfiles)
